src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,309 @@
+        } elseif (is_array($input)) {
+            $this->bson = PackedArray::fromPHP([]);
+            $this->set = array_values($input);
+            $this->exists = array_map(
+                /** @param TValue $value */
+                fn ($value): bool => true,
+                $this->set,
+            );
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [PackedArray::class, 'array', 'null']);
+        }
+
+        $this->codecLibrary = $codecLibrary ?? new LazyBSONCodecLibrary();
+    }
+
+    /** @return array{bson: PackedArray<TValue>, set: array<int, TValue>, unset: array<int, true>, codecLibrary: CodecLibrary} */
+    public function __serialize(): array
+    {
+        return [
+            'bson' => $this->bson,
+            'set' => $this->set,
+            'unset' => $this->unset,
+            'codecLibrary' => $this->codecLibrary,
+        ];
+    }
+
+    /** @param array{bson: PackedArray<TValue>, set: array<int, TValue>, unset: array<int, true>, codecLibrary: CodecLibrary} $data */
+    public function __unserialize(array $data): void
+    {
+        $this->bson = $data['bson'];
+        $this->set = $data['set'];
+        $this->unset = $data['unset'];
+        $this->codecLibrary = $data['codecLibrary'];
+
+        $this->exists = array_map(
+        /** @param TValue $value */
+            fn ($value): bool => true,
+            $this->set,
+        );
+
+        foreach ($this->unset as $index => $unused) {
+            $this->exists[$index] = false;
+        }
+    }
+
+    public function count(): int
+    {
+        $this->readEntirePackedArray();
+
+        return count(array_filter($this->exists));
+    }
+
+    /** @return ListIterator<TValue> */
+    public function getIterator(): ListIterator
+    {
+        $itemIterator = new AppendIterator();
+        // Iterate through all fields in the BSON array
+        $itemIterator->append($this->bson->getIterator());
+        // Then iterate over all fields that were set
+        $itemIterator->append(new ArrayIterator($this->set));
+
+        /** @var array<int, bool> $seen */
+        $seen = [];
+
+        // Use ListIterator to ensure we're indexing from 0 without gaps
+        return new ListIterator(
+            new CallbackIterator(
+                // Skip keys that were unset or handled in a previous iterator
+                new CallbackFilterIterator(
+                    $itemIterator,
+                    /** @param TValue $value */
+                    function ($value, int $offset) use (&$seen): bool {
+                        return ! isset($this->unset[$offset]) && ! isset($seen[$offset]);
+                    },
+                ),
+                /**
+                 * @param TValue $value
+                 * @return TValue
+                 */
+                function ($value, int $offset) use (&$seen) {
+                    // Mark key as seen, skipping any future occurrences
```

Comment:
Yes, the AppendIterator first loops through all elements in the `PackedArray` instance, then over everything that we've set. Any changed value would be present in both, meaning that the `AppendIterator` yields the same field twice. This would cause such elements to be duplicated in the resulting list, as `ListIterator` simply discards the key but does not check if a key was previously seen (iterators can yield the same key multiple times).